usersession/client/client.go

```diff
@@ -0,0 +1,264 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+// dialSessionAgent connects to a user's session agent
+//
+// The host portion of the address is interpreted as the numeric user
+// ID of the target user.
+func dialSessionAgent(network, address string) (net.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	socket := filepath.Join(dirs.XdgRuntimeDirBase, host, "snapd-session-agent.socket")
+	return net.Dial("unix", socket)
+}
+
+type Client struct {
+	doer *http.Client
+}
+
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type response struct {
+	// Not from JSON
+	uid        int
+	err        error
+	statusCode int
+
+	Result json.RawMessage `json:"result"`
+	Type   string          `json:"type"`
+}
+
+func (resp *response) checkError() {
+	if resp.Type != "error" {
+		return
+	}
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) doMany(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
+	sockets, err := filepath.Glob(filepath.Join(dirs.XdgRuntimeDirGlob, "snapd-session-agent.socket"))
+	if err != nil {
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
```

Comment:
I don't think we want logging at this level which means doMany itself should return successes and errors, which then means that Chipaca suggestion might make sense again

Reply:
a solution could also be a callback approach to what to do about this kind of errors

Reply:
to be clear my worry about this logging is also that it might get too noisy and pointless at the same time in some situations

Reply:
I do wonder if it would be better to just silently ignore this case without logging.  This case essentially means that the system administrator has created a non-integer named directory under `/run/user`, and populated it with a file matching the socket path.  It's something that should never be reached, but I agree that it could be noisy the log file if it triggers.

One alternative would be to append another response struct with a dummy value for the UID field (maybe -1?).  But I'm not really sure what the caller could realistically do with that information.

Reply:
yes, after making my initial comment I think this one can be just ignored is a bit too unlikely and the cause would be something local that should be somehow under control anyway